### PR TITLE
Inline SVGTextLayoutAttributes::emptyValue

### PR DESCRIPTION
**Pull request by Ahmad-S792**
<pre>
Inline SVGTextLayoutAttributes::emptyValue
<a href="https://bugs.webkit.org/show_bug.cgi?id=266260">https://bugs.webkit.org/show_bug.cgi?id=266260</a>
rdar://problem/119839011

Reviewed by NOBODY (OOPS!).

Partial Merge: <a href="https://src.chromium.org/viewvc/blink?view=revision&revision=195555">https://src.chromium.org/viewvc/blink?view=revision&revision=195555</a>

This function just returns a simple float - inlining it allows the
compiler to generate more efficient code (and smaller) code
since the relatively high overhead call can be eliminated.

This patch don't merge '"empty value" into NaN' change, since
it leads to regression in one potential test case.

* Source/WebCore/rendering/svg/RenderSVGInlineText.cpp:
(RenderSVGInlineText::characterStartsNewTextChunk):
* Source/WebCore/rendering/svg/SVGTextLayoutAttributes.cpp:
(SVGTextLayoutAttributes::emptyValue): Deleted
* Source/WebCore/rendering/svg/SVGTextLayoutAttributes.h: Inline functions
* Source/WebCore/rendering/svg/SVGTextLayoutAttributesBuilder.cpp:
(SVGTextLayoutAttributesBuilder::buildCharacterDataMap):
(SVGTextLayoutAttributesBuilder::fillCharacterDataMap):
* Source/WebCore/rendering/svg/SVGTextLayoutEngine.cpp:
(SVGTextLayoutEngine::updateCharacterPositionIfNeeded):
(SVGTextLayoutEngine::updateRelativePositionAdjustmentsIfNeeded):
(SVGTextLayoutEngine::layoutTextOnLineOrPath):
</pre>

https://github.com/WebKit/WebKit/commit/cff62739c505be51e7d61e9435ac59d1d775b970

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32058 "Failed to checkout and rebase branch from PR 22031") | [❌ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/10776 "Failed to checkout and rebase branch from PR 22031") | [❌ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/33844 "Failed to checkout and rebase branch from PR 22031") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/34601 "Failed to checkout and rebase branch from PR 22031") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/29053 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [❌ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/32887 "Failed to checkout and rebase branch from PR 22031") | [❌ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/13133 "Failed to checkout and rebase branch from PR 22031") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/51/builds/7999 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/34601 "Failed to checkout and rebase branch from PR 22031") | 
| [❌ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/32457 "Failed to checkout and rebase branch from PR 22031") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/49/builds/13133 "Failed to checkout and rebase branch from PR 22031") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/14/builds/33844 "Failed to checkout and rebase branch from PR 22031") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/34601 "Failed to checkout and rebase branch from PR 22031") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/49/builds/13133 "Failed to checkout and rebase branch from PR 22031") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/14/builds/33844 "Failed to checkout and rebase branch from PR 22031") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/35945 "Failed to checkout and rebase branch from PR 22031") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/49/builds/13133 "Failed to checkout and rebase branch from PR 22031") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/14/builds/33844 "Failed to checkout and rebase branch from PR 22031") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/35945 "Failed to checkout and rebase branch from PR 22031") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/8171 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/51/builds/7999 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/35945 "Failed to checkout and rebase branch from PR 22031") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/9812 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/14/builds/33844 "Failed to checkout and rebase branch from PR 22031") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/8818 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/8696 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | |

### CODE REVIEW

File: Source/WebCore/rendering/svg/SVGTextLayoutAttributes.h [hunk 1]
```diff
@@ -46,7 +48,8 @@ class SVGTextLayoutAttributes {
     explicit SVGTextLayoutAttributes(RenderSVGInlineText&);
 
     void clear();
-    static float emptyValue();
+    static float emptyValue() { return std::numeric_limits<float>::max() - 1; }
+    static bool isEmptyValue(float value) { return std::isnan(value); }
```
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/svg/SVGTextLayoutAttributes.h:52]: That is wrong, you partially merged the NaN-change.

File: Source/WebCore/rendering/svg/SVGTextLayoutAttributes.h [hunk 2]
```diff
@@ -46,7 +48,8 @@ class SVGTextLayoutAttributes {
     explicit SVGTextLayoutAttributes(RenderSVGInlineText&);
 
     void clear();
-    static float emptyValue();
+    static float emptyValue() { return std::numeric_limits<float>::max() - 1; }
```
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/svg/SVGTextLayoutAttributes.h:51]: Nowadays, we should employ constexpr here.

**Review by nikolaszimmermann — CHANGES_REQUESTED** [review]
Needs another iteration.

File: Source/WebCore/rendering/svg/SVGTextLayoutAttributes.h (hunk 1 above)
Comment (Ahmad-S792) [reply Source/WebCore/rendering/svg/SVGTextLayoutAttributes.h:52]: It leads to following test not rendering any text: https://web.archive.org/web/20150712020900/http://output.jsbin.com/imajow/1/quiet

**Pull request closed without merging**